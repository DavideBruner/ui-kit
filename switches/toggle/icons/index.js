import withSizeProp from '../with-size-prop';
import ToggleBackgroundDefaultIcon from './toggle-normal-bkgr.svg';
import ToggleBackgroundHoverIcon from './toggle-active-hover-bkgr.svg';
import ToggleBackgroundDisabledIcon from './toggle-disable-bkgr.svg';
import ToggleBackgroundCheckedIcon from './toggled-bkgr.svg';
import ToggleBackgroundCheckedHoverIcon from './toggled-active-hover-bkgr.svg';
import ToggleBackgroundCheckedDisabledIcon from './toggled-disable-bkgr.svg';
import ToggleButtonDefaultIcon from './toggle-normal-bt.svg';
import ToggleButtonHoverIcon from './toggle-active-hover-bt.svg';
import ToggleButtonDisabledIcon from './toggle-disable-bt.svg';
import ToggleButtonCheckedIcon from './toggled-bt.svg';
import ToggleButtonCheckedHoverIcon from './toggled-active-hover-bt.svg';
import ToggleButtonCheckedDisabledIcon from './toggled-disable-bt.svg';

export default {
  BackgroundDefault: ToggleBackgroundDefaultIcon,
  BackgroundHover: ToggleBackgroundHoverIcon,
  BackgroundDisabled: ToggleBackgroundDisabledIcon,
  BackgroundChecked: ToggleBackgroundCheckedIcon,
  BackgroundCheckedHover: ToggleBackgroundCheckedHoverIcon,
  BackgroundCheckedDisabled: ToggleBackgroundCheckedDisabledIcon,
  ButtonDefault: withSizeProp(ToggleButtonDefaultIcon),
  ButtonHover: withSizeProp(ToggleButtonHoverIcon),
  ButtonDisabled: withSizeProp(ToggleButtonDisabledIcon),
  ButtonChecked: withSizeProp(ToggleButtonCheckedIcon),
  ButtonCheckedHover: withSizeProp(ToggleButtonCheckedHoverIcon),
  ButtonCheckedDisabled: withSizeProp(ToggleButtonCheckedDisabledIcon),
};
