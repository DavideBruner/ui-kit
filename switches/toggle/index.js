import Toggle from './toggle';
import ToggleOn from './toggle-on';
import ToggleOff from './toggle-off';

Toggle.On = ToggleOn;
Toggle.Off = ToggleOff;

export default Toggle;
