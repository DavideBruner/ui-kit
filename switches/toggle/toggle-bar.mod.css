@import '../../materials/spacings.mod.css';
@import '../../materials/colors.mod.css';

:root {
  --big-bar-height: 24px;
  --big-bar-width: 64px;
  --small-bar-height: 12px;
  --small-bar-width: 32px;
}

.toggle-bar {
  position: absolute;
  left: 0;
}

.toggle-bar-small {
  composes: toggle-bar;
  top: calc(var(--spacing-4) / 2);
  width: var(--small-bar-width);
  height: var(--small-bar-height);
}

.toggle-bar-big {
  composes: toggle-bar;
  top: var(--spacing-4);
  width: var(--big-bar-width);
  height: var(--big-bar-height);
}

.toggle-bar rect {
  fill: var(--new-color-gray-60);
}

.toggle-bar.active rect {
  fill: var(--new-color-green-40);
}

.toggle-bar.disabled rect {
  fill: var(--new-color-gray);
}

.toggle-bar.active.disabled rect {
  fill: var(--new-color-green-85);
  opacity: 0.5;
}
