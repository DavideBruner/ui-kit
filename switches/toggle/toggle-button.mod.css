@import '../../materials/colors.mod.css';
@import '../../materials/sizes.mod.css';
@import '../../materials/transitions.mod.css';
@import '../../materials/shadows.mod.css';

:root {
  --big-button-height: var(--standard-input-height);
  --big-button-width: var(--big-button-height);
  --small-button-height: calc(var(--big-button-height) / 2);
  --small-button-width: var(--small-button-height);
}

.toggle-button {
  top: -1px; /* The position of the button was off for some reason */
  left: -1px;
  position: absolute;
  z-index: 1;
  filter: drop-shadow(var(--drop-shadow-1-first))
    drop-shadow(var(--drop-shadow-1-second));
  transition: var(--transition-standard);
}

.toggle-button :global(.bt-fill) {
  transition: var(--transition-standard);
}

.toggle-button-small {
  composes: toggle-button;
  width: var(--small-button-width);
  height: var(--small-button-height);
}

.toggle-button-big {
  composes: toggle-button;
  width: var(--big-button-width);
  height: var(--big-button-height);
}

.toggle-button.active {
  /* The extra 7% shifts the button to the right by about 2px for big and 1px for small */
  transform: translateX(107%);
  filter: drop-shadow(var(--drop-shadow-2));
}

.toggle-button.active :global(.bt-fill) {
  fill: var(--color-buttonPrimary);
}

.toggle-button.disabled {
  filter: drop-shadow(var(--drop-shadow-3));
}

.toggle-button.disabled :global(.bt-fill) {
  fill: var(--color-coolGrey);
}

.toggle-button.disabled.active {
  filter: drop-shadow(var(--drop-shadow-4));
}

.toggle-button.disabled.active :global(.bt-fill) {
  fill: var(--color-greenLight);
}

.toggle-button:hover {
  filter: drop-shadow(var(--drop-shadow-5-first))
    drop-shadow(var(--drop-shadow-5-second));
}

.toggle-button:hover.disabled {
  filter: none;
}
