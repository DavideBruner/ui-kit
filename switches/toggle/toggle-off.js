import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './toggle.mod.css';

const ToggleOff = props =>
  props.isChecked ? null : (
    <span
      className={classnames(styles.label, {
        [styles.textWrapperDisabled]: props.isDisabled,
      })}
    >
      {props.children}
    </span>
  );

ToggleOff.displayName = 'ToggleOff';
ToggleOff.propTypes = {
  isChecked: PropTypes.bool.isRequired,
  isDisabled: PropTypes.bool.isRequired,
  children: PropTypes.node,
};

export default ToggleOff;
