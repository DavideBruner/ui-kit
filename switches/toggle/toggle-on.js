import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './toggle.mod.css';

const ToggleOn = props =>
  props.isChecked ? (
    <span
      className={classnames(styles.toggleLabel, {
        [styles.textWrapperDisabled]: props.isDisabled,
      })}
    >
      {props.children}
    </span>
  ) : null;

ToggleOn.displayName = 'ToggleOn';
ToggleOn.propTypes = {
  isChecked: PropTypes.bool.isRequired,
  isDisabled: PropTypes.bool.isRequired,
  children: PropTypes.node,
};

export default ToggleOn;
