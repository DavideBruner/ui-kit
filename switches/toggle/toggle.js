import React from 'react';
import PropTypes from 'prop-types';
import withMouseOverState from '../../hocs/with-mouse-over-state';
import styles from './toggle.mod.css';
import ToggleSwitch from './toggle-switch';

export class Toggle extends React.PureComponent {
  static displayName = 'Toggle';
  static propTypes = {
    name: PropTypes.string,
    size: PropTypes.string.isRequired,
    isDisabled: PropTypes.bool,
    isChecked: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    children: PropTypes.node,
    // Hoc
    isMouseOver: PropTypes.bool.isRequired,
    handleMouseOver: PropTypes.func.isRequired,
    handleMouseOut: PropTypes.func.isRequired,
  };
  static defaultProps = {
    isDisabled: false,
    isChecked: false,
    size: 'big',
  };

  render() {
    return (
      <div
        onMouseOver={this.props.handleMouseOver}
        onMouseOut={this.props.handleMouseOut}
      >
        <ToggleSwitch
          size={this.props.size}
          isMouseOver={this.props.isMouseOver}
          isChecked={this.props.isChecked}
          isDisabled={this.props.isDisabled}
        />
        <input
          className={styles.inputWrapper}
          name={this.props.name}
          onChange={this.props.onChange}
          disabled={this.props.isDisabled}
          checked={this.props.isChecked}
          type="checkbox"
        />
      </div>
    );
  }
}

export default withMouseOverState(Toggle);
