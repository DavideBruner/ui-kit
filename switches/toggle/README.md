# Switches: Toggle

#### Description

A `Toggle` is an element which indicates an checked or non checked business logic status and
notifies with a value accordingly. It is similar to the `Checkbox` but is styled
like a switch.

#### Usage

```js
import { Toggle } from '@commercetools-local/ui-kit/switches';

<Toggle.Toggle
  isDisabled={false}
  isChecked={false}
  onChange={..}
  size='small'
>
  <Toggle.On>On</Toggle.On>
  <Toggle.Off>Off</Toggle.Off>
</Toggle.Toggle>
```

#### Properties

| Props        | Type     | Required | Values | Default | Description                                         |
| ------------ | -------- | :------: | ------ | ------- | --------------------------------------------------- |
| `children`   | `node`   |    ✅    | -      | -       | The `Toggle.On` and `Toggle.Off` components         |
| `isDisabled` | `bool`   |    -     | -      | `false` | Disables the Toggle                                 |
| `isChecked`  | `bool`   |    -     | -      | `false` | Checks the Toggle                                   |
| `onChange`   | `func`   |    ✅    | -      | -       | What will trigger whenever the `Toggle` is clicked. |
| `size`       | `string` |    ✅    | -      | `big`   | The size of the Toggle component.                   |

#### Invariants

1. The `Toggle.Toggle` must have one each of the `Toggle.On` and `Toggle.Off` components.
