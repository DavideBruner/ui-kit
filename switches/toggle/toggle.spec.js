import React from 'react';
import { shallow } from 'enzyme';
import { Toggle } from './toggle';
import ToggleOn from './toggle-on';
import ToggleOff from './toggle-off';

const onChangeCallback = jest.fn();

const createTestProps = custom => ({
  name: 'toggle',
  isDisabled: false,
  isChecked: false,
  onChange: onChangeCallback,

  // HoC
  handleMouseOver: jest.fn(),
  handleMouseOut: jest.fn(),
  isMouseOver: false,
  ...custom,
});

describe('<Toggle>', () => {
  describe('rendering', () => {
    let props;
    let wrapper;
    describe('without children', () => {
      beforeEach(() => {
        props = createTestProps();

        wrapper = shallow(<Toggle {...props} />);
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should supply `onChange` to the input', () => {
        expect(wrapper.find('input')).toHaveProp('onChange', onChangeCallback);
      });

      describe('when disabled', () => {
        beforeEach(() => {
          props = createTestProps({ isDisabled: true });

          wrapper = shallow(<Toggle {...props} />);
        });

        it('should disable the `input`', () => {
          expect(wrapper.find('input')).toHaveProp('disabled', true);
        });
      });

      describe('when checked', () => {
        beforeEach(() => {
          props = createTestProps({ isChecked: true });

          wrapper = shallow(<Toggle {...props} />);
        });

        it('should check the `input`', () => {
          expect(wrapper.find('input')).toHaveProp('checked', true);
        });
      });
    });

    describe('with children', () => {
      beforeEach(() => {
        props = createTestProps();

        wrapper = shallow(
          <Toggle>
            <ToggleOn>{'On'}</ToggleOn>
            <ToggleOff>{'Off'}</ToggleOff>
          </Toggle>
        );
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe('interacting', () => {
    describe('onChange', () => {
      let props;
      beforeEach(() => {
        props = createTestProps({ isChecked: true });

        props.onChange();
      });

      it('should invoke the `onChange` callback', () => {
        expect(onChangeCallback).toHaveBeenCalled();
      });
    });
  });
});
