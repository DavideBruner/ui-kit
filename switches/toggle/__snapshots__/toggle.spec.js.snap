// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle> rendering with children should match snapshot 1`] = `
<label
  className="labelWrapper"
  onMouseOut={[MockFunction]}
  onMouseOver={[MockFunction]}
>
  <ToggleSwitch
    isChecked={false}
    isDisabled={false}
    isMouseOver={false}
    size="big"
  />
  <input
    checked={false}
    className="inputWrapper"
    disabled={false}
    name="toggle"
    onChange={[Function]}
    type="checkbox"
  />
</label>
`;

exports[`<Toggle> rendering without children should match snapshot 1`] = `
<label
  className="labelWrapper"
  onMouseOut={[MockFunction]}
  onMouseOver={[MockFunction]}
>
  <ToggleSwitch
    isChecked={false}
    isDisabled={false}
    isMouseOver={false}
    size="big"
  />
  <input
    checked={false}
    className="inputWrapper"
    disabled={false}
    name="toggle"
    onChange={[Function]}
    type="checkbox"
  />
</label>
`;
